packages/modules/event-bus-redis/src/services/event-bus-redis.ts

```diff
@@ -71,78 +81,114 @@ export default class RedisEventBusService extends AbstractEventBusModuleService
-    options: BulkJobOptions | JobsOptions = {}
-  ): Promise<void> {
-    const globalJobOptions = this.moduleOptions_.jobOptions ?? {}
+    const eventsToGroup = eventsDataArray.filter((eventData) =>
+      isPresent(eventData.body?.metadata?.eventGroupId)
+    )
 
-    const isBulkEmit = Array.isArray(eventNameOrData)
+    const groupEventsMap = new Map<string, Message<T>[]>()
 
-    const opts = {
-      // default options
-      removeOnComplete: true,
-      attempts: 1,
-      // global options
-      ...globalJobOptions,
-    } as EmitOptions
-
-    const dataBody = isString(eventNameOrData)
-      ? data ?? (data as Message<T>).body
-      : undefined
-
-    const events = isBulkEmit
-      ? eventNameOrData.map((event) => ({
-          name: event.eventName,
-          data: {
-            eventName: event.eventName,
-            data: (event as EmitData).data ?? (event as Message<T>).body,
-          },
-          opts: {
-            ...opts,
-            // local options
-            ...event.options,
-          },
-        }))
-      : [
-          {
-            name: eventNameOrData as string,
-            data: { eventName: eventNameOrData, data: dataBody },
-            opts: {
-              ...opts,
-              // local options
-              ...options,
-            },
-          },
-        ]
-
-    await this.queue_.addBulk(events)
+    for (const event of eventsToGroup) {
+      const groupId = event.body.metadata?.eventGroupId!
+      const array = groupEventsMap.get(groupId) ?? []
+
+      array.push(event)
+      groupEventsMap.set(groupId, array)
+    }
+
+    const promises: Promise<unknown>[] = []
+
+    if (eventsToEmit.length) {
+      const emitData = this.buildEvents(eventsToEmit, options)
+
+      promises.push(this.queue_.addBulk(emitData))
+    }
+
+    for (const [groupId, events] of groupEventsMap.entries()) {
+      const eventsData = this.buildEvents(events, options)
+
+      promises.push(this.groupEvents(groupId, eventsData))
+    }
+
+    await promiseAll(promises)
   }
 
-  // TODO: Implement redis based staging + release
-  async releaseGroupedEvents(eventGroupId: string) {}
-  async clearGroupedEvents(eventGroupId: string) {}
+  private async groupEvents<T = unknown>(
+    eventGroupId: string,
+    events: IORedisEventType<T>[]
+  ) {
+    await this.eventBusRedisConnection_.rpush(
+      `staging:${eventGroupId}`,
+      ...events.map((event) => JSON.stringify(event))
+    )
+  }
+
+  private async getGroupedEvents(eventGroupId: string) {
+    return await this.eventBusRedisConnection_
+      .lrange(`staging:${eventGroupId}`, 0, -1)
+      .then((result) => {
+        return result.map((jsonString) => JSON.parse(jsonString))
+      })
+  }
+
+  async releaseGroupedEvents(eventGroupId: string) {
+    const groupedEvents: IORedisEventType[] = await this.getGroupedEvents(
+      eventGroupId
+    )
+
+    await this.queue_.addBulk(groupedEvents)
+
+    await this.clearGroupedEvents(eventGroupId)
```

Comment:
We might need a lock in this method to avoid concurrency. Also, we will probably want to have a flag that is set before we start adding new events to the main queue, as we could easily get the events, emit them, but then the clearing fails, which will probably result in duplicate events.

Reply:
maybe we should grab, clear and emit instead of grab, emit and clear. could this work?

Reply:
If we can have all these operations in a Lua script we reduce a lot the points of failure. I believe for this PR this is ok, but ideally we try to do all that in one single call and hope Redis won't fail in the middle of the operation. That will be the single point of failure.
We just need a script that sets all the messages in the format bullmq is expecting.

Reply:
we have indeed a separete ticket for that

Reply:
grab, clear, and emit is a bit worse, since you might lose messages (at least once delivery is almost always better than at most once). Doing it all in Redis would increase the reliability for sure, I am just thinking if it will cause issues with eg. bullmq changing something under the surface (but not the public API) and our solution breaking. But we can definitely discuss it, I just wanted to raise it here so at the very least we add a TODO or a ticket.

Reply:
we probably need to write a test to read all the keys created by bullmq, at least if we bump the version and it has changed we know.

Reply:
I created a separate ticket for the lua scripts because I think that will open up some questions of its own :D 

Wrt to the fail in clearing events, i think we can remove clearing events from this function and pass that responsibility onto the consumer to handle the failure scenario. If the consumer ends up not clearing them, the default TTL should clear it up eventually

wdyt?

Reply:
We can just ignore or add a log if clearing events fail. THe worst case scenario will be garbage left in Redis.
BullMQ won't manage this separeted queue, so nothing will be consumed.

Reply:
@riqwan I think just noting it down in comments as a possibility is enough for now. 

The main problem is you don't know if the events were emitted before clearing or not, so you don't know if you should retry the operation. It doesn't really matter who does the clearing, the issue is if a machine crashes, you don't know where to proceed from without a flag (similar to workflows).